src/meta/src/stream/source_manager.rs

```diff
@@ -972,13 +1011,89 @@ impl SourceManager {
         Ok(assigned)
     }
 
+    /// Allocates splits to actors for replace source job.
+    pub async fn allocate_splits_for_replace_source(
+        &self,
+        job_id: &TableId,
+        merge_updates: &Vec<MergeUpdate>,
+    ) -> MetaResult<SplitAssignment> {
+        tracing::debug!(?merge_updates, "allocate_splits_for_replace_source");
+        if merge_updates.is_empty() {
+            // no existing downstream. We can just re-allocate splits arbitrarily.
+            return self.allocate_splits(job_id).await;
+        }
+
+        let core = self.core.lock().await;
+        let table_fragments = core
+            .metadata_manager
+            .get_job_fragments_by_id(job_id)
+            .await?;
+
+        let source_fragments = table_fragments.stream_source_fragments();
+        assert_eq!(
+            source_fragments.len(),
+            1,
+            "replace source job should only have one source"
+        );
+        let (_source_id, fragments) = source_fragments.into_iter().next().unwrap();
+        assert_eq!(
+            fragments.len(),
+            1,
+            "replace source job should only have one fragment"
+        );
+        let fragment_id = fragments.into_iter().next().unwrap();
+
+        debug_assert!(
+            !merge_updates.is_empty()
+                && merge_updates.iter().all(|merge_update| {
+                    merge_update.upstream_fragment_id == merge_updates[0].upstream_fragment_id
+                        && merge_update.new_upstream_fragment_id == Some(fragment_id)
+                }),
+            "merge update should only replace one fragment: {:?}",
+            merge_updates
+        );
+        let prev_fragment_id = merge_updates[0].upstream_fragment_id;
+        // Here we align the new source executor to backfill executors
+        //
+        // old_source           backfill_1            new_source
+        // actor_x1-----┬------>actor_a1 -----┬------>actor_y1
+        // actor_x2-----┼-┬---->actor_a2 -----┼-┬---->actor_y2
+        //              │ │                   │ │
+        //              │ │     backfill_2    │ │
+        //              └-┼---->actor_b1------┘ │
+        //                └---->actor_b2--------┘
+        //
+        // Note: we can choose any backfill actor to align here.
+        // We use `HashMap` to dedup.
```

Comment:
I don't quite get why backfill is appearing here.

Reply:
- When creating SourceBackfill executors, we need to align split assignment to the source executor. (i.e., actors connected with NoShuffle should have the same splits)
- Therefore, the real goal is that when replacing source executors, we need to ensure the new executors still have the aligned splits.

Reply:
The graph might be a little misleading.